bin/runtime-common/src/messages_xcm_extension.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+// This file is part of Cumulus.
+
+// Cumulus is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Cumulus is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Module provides utilities for easier XCM handling, e.g:
+//! [`XcmExecutor`] -> [`MessageSender`] -> <outbound message queue>
+//!												|
+//!											<relayer>
+//!												|
+//! [`XcmRouter`] <- [`MessageDispatch`] <- <inbound message queue>
+
+use bp_messages::{
+	source_chain::MessagesBridge,
+	target_chain::{DispatchMessage, MessageDispatch},
+	LaneId,
+};
+use bp_runtime::{messages::MessageDispatchResult, AccountIdOf, Chain};
+use codec::{Decode, Encode};
+use frame_support::{dispatch::Weight, CloneNoBound, EqNoBound, PartialEqNoBound};
+use scale_info::TypeInfo;
+use xcm_builder::{DispatchBlob, DispatchBlobError, HaulBlob, HaulBlobError};
+
+/// PLain "XCM" payload, which we transfer through bridge
+pub type XcmAsPlainPayload = sp_std::prelude::Vec<u8>;
+
+/// Message dispatch result type for single message
+#[derive(CloneNoBound, EqNoBound, PartialEqNoBound, Encode, Decode, Debug, TypeInfo)]
+pub enum XcmBlobMessageDispatchResult {
+	InvalidPayload,
+	Dispatched,
+	NotDispatched(#[codec(skip)] &'static str),
+}
+
+/// [`XcmBlobMessageDispatch`] is responsible for dispatching received messages
+pub struct XcmBlobMessageDispatch<SourceBridgeHubChain, TargetBridgeHubChain, DispatchBlob> {
```

Comment:
well, yes, `FromBridgedChainMessageDispatch` is similar but uses our struct `FromBridgedChainMessagePayload`, 
`XcmBlobMessageDispatch` plays with xcm blob and does not even decode xcm, but also does not touches `dispatch_weight` and weights/fees for bridge-hub are other issue (coming soon)

so I would not touch `FromBridgedChainMessageDispatch` now, unless we set up weight/fees for BridgeHub, then we will see better

Reply:
anyway, I thought that could help to fix those integrity_tests, but it is unrelated. If we could change integrity test that InboundPayload could take `FromBridgedChainMessagePayload` or `Vec<u8>`.

all this stuff, it is just simple connection point between XcmExecutor/XcmConfig handling ExportMessage and messages pallet (outbound) queue on one side and between message pallet (inbound) queue triggering dispatch with XcmRouter on the other side, so all this is really just xcm extension, because rialto/millau setup does not use `xcm-v3::ExportMessage`

we dont need to even backport it now, I can still have it in Cumulus, but it looks like it could fit to bridges repo also, so later when weight/fees will be stable we could move it here

Reply:
I agree that we don't need `FromBridgedChainMessageDispatch` anymore (dispatch is not a goal of this repo anymore). But we can do it in a separate PR to avoid delays in RBH<>WBH implementation (I can do that myself).

@bkontur This `dispatch_weight` method is a weight of message dispatch at **this chain**. So if we are talking about bridge hubs, which are just redirecting messages to other parachains, this should be the weight of this **redirection**, not the weight of the XCM message itself (which happens at other parachain). So e.g. if redirecting is just appending it to some map/queue, it may be just `DbWegith::read_writes(1, 1)` or something like that. In any case - you don't need to use `XcmExecutor` or decode XCM message here (unless I'm totally wrong).

Reply:
@svyatonik 
ok, thank you, I think I understand now this weight stuff more and more,

I added here `DispatchBlobWeigher: Get<Weight>`, so the runtime will provide some values,

e.g. BridgeHub for dispatch uses:
```
pub type XcmRouter = (
	// Two routers - use UMP to communicate with the relay chain:
	cumulus_primitives_utility::ParentAsUmp<ParachainSystem, PolkadotXcm, ()>,
	// ..and XCMP to communicate with the sibling chains.
	XcmpQueue,
);
```
`XcmpQueue` does cca 2 reads and 2 writes, so it will be `DbWegith::read_writes(2, 2)`